ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/channels/RangedSeekableByteChannel.java

```diff
@@ -0,0 +1,196 @@
+package com.spectralogic.ds3client.helpers.channels;
+
+import com.google.common.collect.*;
+import com.spectralogic.ds3client.helpers.TruncateNotAllowedException;
+import com.spectralogic.ds3client.models.Range;
+import com.spectralogic.ds3client.models.bulk.BulkObject;
+import com.spectralogic.ds3client.utils.Guard;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.SeekableByteChannel;
+import java.util.HashMap;
+import java.util.Map;
+
+public class RangedSeekableByteChannel implements SeekableByteChannel {
+
+    private final SeekableByteChannel byteChannel;
+    private final ImmutableMultimap<BulkObject, Range> ranges;
+    private final ImmutableMap<BulkObject, Long> blobSizes;
+    private final ImmutableMap<BulkObject, Long> startingOffsetForBlob;
+    private final long size;
+
+    private long position;
+    private boolean open;
+
+    public static RangedSeekableByteChannel wrap(final SeekableByteChannel byteChannel, final ImmutableMultimap<BulkObject, Range> ranges) throws IOException {
+        return new RangedSeekableByteChannel(byteChannel, ranges);
+    }
+
+    public RangedSeekableByteChannel(final SeekableByteChannel byteChannel, final ImmutableMultimap<BulkObject, Range> ranges) throws IOException {
+        this.byteChannel = byteChannel;
+        this.ranges = ranges;
+        this.position = 0;
+        this.open = true;
+        this.size = getSize(byteChannel.size(), ranges);
+        this.blobSizes = computesBlobSize(ranges);
+        this.startingOffsetForBlob = computeRealBlobOffset(this.blobSizes);
+    }
+
+    private static ImmutableMap<BulkObject,Long> computeRealBlobOffset(final ImmutableMap<BulkObject, Long> blobLengths) {
+        if (Guard.isMapNullOrEmpty(blobLengths)) {
+            return ImmutableMap.of();
+        }
+        final Map<BulkObject, Long> realOffsets = new HashMap<>();
+
+        final ImmutableSortedSet<BulkObject> bulkObjects = ImmutableSortedSet.copyOf(BlobComparator.create(), blobLengths.keySet());
+        final ImmutableList<BulkObject> sortedList = ImmutableList.copyOf(bulkObjects);
+
+        final int listLength = sortedList.size();
+
+        for(int i = 0; i < listLength; i++) {
+            if(i == 0) {
+                realOffsets.put(sortedList.get(0), 0L);
```

Comment:
If it were before the loop I would have to check to make sure that `listLength != 0`.  It probably wouldn't be hit because of the Guard statement, but this is why.